reverse_engineering/helpers/documentKindHelper.js

```diff
@@ -0,0 +1,299 @@
+			const errorMessage = clusterHelper.getErrorMessage({ error });
+			logger.info(keySpaceName + errorMessage);
+			return getDefaultDocumentKindData({ bucketName, status: STATUS.hasError });
+		});
+};
+
+/**
+ * @param {{ bucketName: string; documents: Document[]; }} param0
+ * @returns {object}
+ */
+const generateCustomInferSchema = ({ bucketName, documents }) => {
+	documents = documents.map(item => {
+		if (typeof item[bucketName] === 'object') {
+			return item[bucketName];
+		} else {
+			return { 'unknown': item[bucketName] };
+		}
+	});
+
+	let inferSchema = {
+		'#docs': 0,
+		'$schema': 'http://json-schema.org/schema#',
+		'properties': {},
+	};
+
+	documents.forEach(item => {
+		inferSchema['#docs']++;
+
+		for (const prop in item) {
+			if (inferSchema.properties.hasOwnProperty(prop)) {
+				const minimumNumberOfSamples = 20;
+				const samples = inferSchema.properties[prop]['samples'];
+				const propDoesNotExists = samples.indexOf(item[prop]) === -1;
+				const shouldAddSample = samples.length < minimumNumberOfSamples;
+
+				inferSchema.properties[prop]['#docs']++;
+
+				if (propDoesNotExists && shouldAddSample) {
+					inferSchema.properties[prop]['samples'].push(item[prop]);
+				}
+
+				inferSchema.properties[prop]['type'] = schemaHelper.typeOf(item[prop]);
+			} else {
+				inferSchema.properties[prop] = {
+					'#docs': 1,
+					'%docs': 100,
+					'samples': [item[prop]],
+					'type': schemaHelper.typeOf(item[prop]),
+				};
+			}
+		}
+	});
+
+	for (const prop in inferSchema.properties) {
+		inferSchema.properties[prop]['%docs'] = _.round(
+			(inferSchema.properties[prop]['#docs'] / inferSchema['#docs']) * 100,
+			2,
+		);
+	}
+	return inferSchema;
+};
+
+/**
+ * @param {Document} property
+ * @returns {boolean}
+ */
+const isSuggestedDocKind = property => {
+	return property['%docs'] >= 70 && property.samples?.length && typeof property.samples[0] !== 'object';
+};
+
+/**
+ * @param {string[]} array
+ * @returns {string[]}
+ */
+const replaceQuotes = array =>
+	array.map(item => {
+		const inQuotes = _.first(item) === '`' && _.last(item) === '`';
+
+		return inQuotes ? item.slice(1, -1) : item;
+	});
+
+const rejectPropertiesWithLowAppearancePercentage = properties => {
+	return _.chain(properties)
+		.toPairs()
+		.filter(([, propertyData]) => isSuggestedDocKind(propertyData))
+		.fromPairs()
+		.value();
+};
+
+/**
+ * @param {Document[]} inferences
+ * @returns {string[]}
+ */
+const getSuggestedDocKindsFromInference = inferences => {
+	return _.uniq(
+		inferences.flatMap(inference => Object.keys(rejectPropertiesWithLowAppearancePercentage(inference.properties))),
+	);
+};
+
+/**
+ * @param {{ bucketName: string; inference: Document; isCustomInfer: boolean; flavorValue: string; }} param0
+ * @returns {DocumentKindData}
+ */
+const getDocumentKindDataFromInfer = ({ bucketName, inference, isCustomInfer, flavorValue }) => {
+	let suggestedDocKinds = [];
+	let otherDocKinds = [];
+	let documentKind = {
```

Comment:
is there any documented agreement regarding that we declare a mutable variable content by `let` instead of `const`, when we are not changing the var ref? Just curious...